lib/garbage-mpu-parts-cleaner.js

```diff
@@ -0,0 +1,607 @@
+
+    req.once('error', function _findError(err) {
+        self.log.error(
+            {
+                record: record,
+                elapsed: elapsedSince(listBegin),
+                err: err
+            },
+            'Error reading MPU upload data.'
+        );
+
+        self.addCounter('partListErrors', 1);
+
+        if (common.isMorayOverloaded(err)) {
+            // TODO: automatically back off
+            self.log.warn('Warning Moray is overloaded, we should back off.');
+        }
+
+        callback(err);
+    });
+
+    req.on('record', function _findRecord(obj) {
+        var value = obj.value;
+
+        deleteBatch.push({
+            operation: 'delete',
+            bucket: common.MANTA_BUCKET,
+            key: obj.key,
+            options: {
+                etag: obj._etag,
+                // These "headers" are required in order for the delete to
+                // trigger a manta_fastdelete_queue entry.
+                headers: {
+                    'x-muskie-prev-metadata': value,
+                    'x-muskie-snaplinks-disabled': true
+                }
+            }
+        });
+    });
+
+    req.once('end', function _findEnd() {
+        var deleteBegin = process.hrtime();
+        var elapsed = elapsedSince(listBegin);
+        var reqId = uuid();
+
+        self.log.debug(
+            {
+                deleteBatch: deleteBatch,
+                elapsed: elapsed
+            },
+            'Finished finding MPU upload parts.'
+        );
+
+        assert.ok(
+            deleteBatch.length <= MAX_PARTS,
+            '%d parts should be enough for everyone. %d is too many.',
+            MAX_PARTS,
+            deleteBatch.length
+        );
+
+        self.addCounter('partListSeconds', elapsed);
+
+        if (deleteBatch.length === 0) {
+            self.log.info(
+                {
+                    uploadDir: uploadDirname
+                },
+                'No parts objects found, no need for batch delete.'
+            );
+
+            callback(null, {
+                deletedParts: deletedParts
+            });
+            return;
+        } else {
+            self.addCounter('partListRecordCount', deleteBatch.length);
+            self.addCounter('partDeleteBatchAttempts', 1);
+
+            self.client.batch(deleteBatch, {req_id: reqId}, function _didBatch(
+                err,
+                res
+            ) {
+                var logLevel = 'debug';
+
+                elapsed = elapsedSince(deleteBegin);
+
+                self.addCounter('partDeleteBatchSeconds', elapsed);
+
+                if (err) {
+                    self.addCounter('partDeleteBatchErrors', 1);
+                    logLevel = 'error';
+                } else {
+                    assert.array(res.etags, 'res.etags');
+                    self.addCounter('partDeleteSuccessCount', res.etags.length);
+                    deletedParts = res.etags.length;
```

Comment:
XXX What does it mean if res.etags.length is less than deleteBatch.length?

Reply:
I'll add a comment about that. And an assert.